test/verify/check-login

```diff
@@ -561,5 +561,135 @@ account    required     pam_succeed_if.so user ingroup %s""" % m.get_admin_group
     def testFailingWebsocketSafariNoCA(self):
         self.testFailingWebsocket(safari=True, cacert=False)
 
+    def testTally(self):
+        m = self.machine
+        b = self.browser
+
+        if m.image.startswith('debian') or m.image.startswith('ubuntu'):
+            module = 'pam_tally2'
+            logfile = lambda user: '/var/log/tallylog'
+        elif m.image.startswith('rhel') or m.image.startswith('fedora') or m.image.startswith('cent'):
+            module = 'faillock'
+            logfile = lambda user: '/var/run/faillock/' + user
+        else:
+            raise ValueError('unexpected image name', m.image)
+
+        def expect_fail_count(expected, user="admin", must_exist=True):
+            if must_exist:
+                cksum = m.execute("cksum {}".format(logfile(user)))
+
+            output = m.execute("{} --user {}".format(module, user))
+
+            if module == 'faillock':
+                n_lines = len(output.splitlines())
+                if n_lines:
+                    n_lines -= 2 # remove the header, if it's there
+
+                self.assertEqual(expected, n_lines)
+
+            else: # pam_tally2
+                self.assertIn(' {} '.format(expected), output)
+
+            # make sure the above command didn't change the file
+            if must_exist:
+                self.assertEqual(cksum, m.execute("cksum {}".format(logfile(user))))
+
+        def expect_failed_login(user, password, n_failed):
+            b.try_login(user, password)
+            b.wait_text_not("#login-error-message", "")
+            expect_fail_count(n_failed, user=user)
+
+        def expect_successful_login(user, password):
+            b.login_and_go('/system', user=user, password=password)
+            expect_fail_count(0, user=user)
+            b.logout()
+
+        # ensure we have no module in our pam config already
+        m.execute("! grep -r '{}' /etc/pam.d".format(module))
+
+        # add it to the pam config
+        if module == 'pam_tally2':
+            self.restore_file("/etc/pam.d/cockpit")
+            m.execute("sed -i '/common-auth/ {"
+                      "iauth required pam_tally2.so deny=4\n"
+                      "}' /etc/pam.d/cockpit")
+        else:
+            self.restore_file("/etc/pam.d/password-auth")
+            m.execute("sed -i '/pam_unix/ {"
```

Comment:
These commands are quite useful as a reference how to set this up. Did you take them from some official documentation? Then it would be nice to add comments with the URLs for these (like https://access.redhat.com/solutions/62949)

Reply:
I took the faillock ones from the page you sent me, but wrote the sed hackery myself.